lib/membrane_rtmp_plugin/rtmp/sink/sink.ex

```diff
@@ -114,70 +128,63 @@ defmodule Membrane.RTMP.Sink do
-           buffer.dts,
-           buffer.metadata.h264.key_frame?
-         ) do
-      {:ok, native} ->
-        state = Map.put(state, :native, native)
-
-        state =
-          Map.update(state, :current_timestamps, %{video: buffer.dts}, fn curr_tmps ->
-            Map.put(curr_tmps, :video, buffer.dts)
-          end)
-
-        {{:ok, demand: state.current_timestamps |> Enum.min_by(&elem(&1, 1)) |> elem(0)}, state}
+  def handle_write(pad, buffer, _ctx, state) do
+    state = write_frame(state, pad, buffer)
+    {{:ok, demand: get_demand(state)}, state}
+  end
 
-      {:error, reason} ->
-        raise("Writing video frame failed with reason: #{reason}")
+  @impl true
+  def handle_end_of_stream(pad, ctx, state) do
+    if ctx.pads |> Map.values() |> Enum.map(& &1.end_of_stream?) |> Enum.all?() do
+      {:ok, state}
+    else
+      state = Map.update!(state, :current_timestamps, &Map.delete(&1, pad))
+      {{:ok, demand: get_demand(state)}, state}
     end
   end
 
-  @impl true
-  def handle_write(:audio, buffer, _ctx, %{ready: true} = state) do
+  defp write_frame(state, :audio, buffer) do
     buffer_pts = buffer.pts |> Ratio.ceil()
 
     case Native.write_audio_frame(state.native, buffer.payload, buffer_pts) do
       {:ok, native} ->
-        state = Map.put(state, :native, native)
-
-        state =
-          Map.update(state, :current_timestamps, %{audio: buffer_pts}, fn curr_tmps ->
-            Map.put(curr_tmps, :audio, buffer_pts)
-          end)
-
-        {{:ok, demand: state.current_timestamps |> Enum.min_by(&elem(&1, 1)) |> elem(0)}, state}
+        Map.put(state, :native, native)
+        |> Map.update!(:current_timestamps, fn curr_tmps ->
+          Map.put(curr_tmps, :audio, buffer_pts)
+        end)
 
       {:error, reason} ->
         raise("Writing audio frame failed with reason: #{reason}")
     end
   end
 
-  @impl true
-  def handle_end_of_stream(
-        pad,
-        _ctx,
-        %{current_timestamps: %{audio: _audio_dts, video: _video_dts} = curr_tmps} = state
-      ) do
-    curr_tmps = Map.delete(curr_tmps, pad)
-    state = Map.put(state, :current_timestamps, curr_tmps)
-    {{:ok, demand: curr_tmps |> Map.keys() |> List.first()}, state}
+  defp write_frame(state, :video, buffer) do
+    case Native.write_video_frame(
+           state.native,
+           buffer.payload,
+           buffer.dts,
+           buffer.metadata.h264.key_frame?
+         ) do
+      {:ok, native} ->
+        Map.put(state, :native, native)
+        |> Map.update!(:current_timestamps, fn curr_tmps ->
+          Map.put(curr_tmps, :video, buffer.dts)
+        end)
+
+      {:error, reason} ->
+        raise("Writing video frame failed with reason: #{reason}")
+    end
   end
 
-  @impl true
-  def handle_end_of_stream(_pad, _ctx, state) do
-    {:ok, Map.delete(state, :current_timestamps)}
+  defp get_demand(state) do
+    state.current_timestamps |> Enum.min_by(fn {_pad, timestamp} -> timestamp end) |> elem(0)
```

Comment:
Three things:
- avoid `elem/2` if possible
- timestamp can be `Ratio.t()`, and then it will be constantly demanding from one pad - you need to fix that. Also you can use `Bunch.value/1`, but I don't insist
- I'd rather you returned `{pad, 1}`. This will signal the intention a little more clearly

Reply:
I make sure that timestamp of audio frame is integer at line 155 in sink.ex.